runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java

```diff
@@ -659,83 +754,196 @@ private void setCurrentOutputWatermark(long currentOutputWatermark) {
   }
 
   /**
-   * Factory for creating an {@link DoFnRunners.OutputManager} from
+   * Factory for creating an {@link BufferedOutputManager} from
    * a Flink {@link Output}.
    */
   interface OutputManagerFactory<OutputT> extends Serializable {
-    DoFnRunners.OutputManager create(Output<StreamRecord<WindowedValue<OutputT>>> output);
+    BufferedOutputManager<OutputT> create(
+        Output<StreamRecord<WindowedValue<OutputT>>> output,
+        StateInternals stateInternals);
   }
 
   /**
-   * Default implementation of {@link OutputManagerFactory} that creates an
-   * {@link DoFnRunners.OutputManager} that only writes to
-   * a single logical output.
+   * A {@link DoFnRunners.OutputManager} that can buffer its outputs.
+   * Use {@link FlinkSplitStateInternals} or {@link FlinkKeyGroupStateInternals}
+   * to keep buffer data.
    */
-  public static class DefaultOutputManagerFactory<OutputT>
-      implements OutputManagerFactory<OutputT> {
-    @Override
-    public DoFnRunners.OutputManager create(
-        final Output<StreamRecord<WindowedValue<OutputT>>> output) {
-      return new DoFnRunners.OutputManager() {
-        @Override
-        public <T> void output(TupleTag<T> tag, WindowedValue<T> value) {
-          // with tagged outputs we can't get around this because we don't
-          // know our own output type...
-          @SuppressWarnings("unchecked")
-          WindowedValue<OutputT> castValue = (WindowedValue<OutputT>) value;
-          output.collect(new StreamRecord<>(castValue));
+  public static class BufferedOutputManager<OutputT> implements
+      DoFnRunners.OutputManager {
+
+    private final BiMap<TupleTag<?>, Integer> tagToLabel;
+    private TupleTag<OutputT> mainTag;
+    private Map<TupleTag<?>, OutputTag<WindowedValue<?>>> mapping;
+    protected Output<StreamRecord<WindowedValue<OutputT>>> output;
+
+    private boolean openBuffer = false;
+    private BagState<KV<Integer, WindowedValue<?>>> bufferState;
+
+    BufferedOutputManager(
+        Output<StreamRecord<WindowedValue<OutputT>>> output,
+        TupleTag<OutputT> mainTag,
+        Map<TupleTag<?>, OutputTag<WindowedValue<?>>> mapping,
+        final Map<TupleTag<?>, Coder<WindowedValue<?>>> coderMapping,
+        StateInternals stateInternals) {
+      this.output = output;
+      this.mainTag = mainTag;
+      this.mapping = mapping;
+
+      tagToLabel = transformTupleTagsToLabels(mainTag, mapping);
+
+      StateTag<BagState<KV<Integer, WindowedValue<?>>>> bufferTag =
```

Comment:
I just realised that storing elements in state with some label or other derived from the `TupleTag` is problematic if they user doesn't explicitly specify a name for the tag. For example, when a program is restored from a savepoint, the code that generates the program graph will be executed again and the (generated) names of tags will be different.

Also, assigning a `TupleTag` id as in `transformTupleTagsToLabels()` must be done in some deterministic way, for example by sorting tags based on the name. Otherwise we will get the same problem when restoring.

@kennknowles How does the Dataflow Runner handle this, or do you never store tags there? OR, should tags never be persisted in the first place? (I'm getting the feeling that they shouldn't.)